### PR TITLE
feat(ci): Balance test distribution for pytest

### PR DESCRIPTION
**Pull request by billyvg**
Currently, we distribute the tests based only on filename. This PR will save pytest results in GH as an artifact. The workflows will then download these results before running the tests and distribute tests based on how long they took to run. The test collection process works as follows:

* iterate through each test case
* find the test group that has the lowest total test duration assigned to it
* assign test case to be run on that test group
* if that test case is found in the pytest results, update the test group duration

This should work pretty ok as we are distributing individual test cases (vs test file), as I could see problems if didn't have an even distribution of test durations (e.g. many slow tests + many fast tests and not much in between, which is more likely to happen if we could only distribute via files).

Ideally this would also be an external service where we would average out the results as well instead of only using the latest.

### CODE REVIEW

File: .github/workflows/backend-test.yml [hunk 1]
```diff
@@ -24,12 +24,19 @@ jobs:
...
           # https://github.com/codecov/codecov-bash/blob/7100762afbc822b91806a6574658129fe0d23a7d/codecov#L891
           fetch-depth: '2'
 
+      - name: Download a previous pytest report
+        continue-on-error: true
+        uses: ./.github/actions/pytest-report
+        with:
+          action: 'download'
+          # testing
+          branch: 'feat/ci/balance-pytest-test-durations'
```
Comment (billyvg) [inline .github/workflows/backend-test.yml:38]: This will be removed as it should fetch from main branch.

**Comment by billyvg** [comment]
Some runs work better than others:

https://github.com/getsentry/sentry/runs/4822012995?check_suite_focus=true (here they're all around 9 minutes).

File: .github/actions/pytest-report/action.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        branch: ${{ inputs.branch }}
+        artifact: ${{ inputs.artifact }}
+        path: ${{ github.workspace }}/${{ inputs.path }}
+
+    - name: Debug
+      if: inputs.action != 'upload'
+      shell: bash
+      run: |
+        ls ${{ github.workspace }}/${{ inputs.path }}
+        echo $(pwd)
```
Comment (joshuarli) [inline .github/actions/pytest-report/action.yml:44]: nit: You can actually just access it with `$PWD`. You don't need to subshell.

**Review by joshuarli — APPROVED** [review]

**Review by joshuarli — CHANGES_REQUESTED** [review]
@billyvg Actually, a concern: it would be annoying if tests shifted around from one job to the next. Like if you fail various tests, and then are looking for the next commit's result for a specific test, it would take more time to find if it got rebalanced to a new job. I think this is definitely going to happen for the tests with retries.

You probably want to hash the branch name and use that to freeze the distribution per PR.

**Comment by billyvg** [comment]
@joshuarli That's a good point, I wonder if that's how users would approach it though? If you're fixing tests it seems likely that you'll look for failed jobs rather than a specific check?

Another thing that's potentially annoying is that this distributes individual tests vs files, so a file can be run on all instances and makes it more difficult to track down.

**Comment by joshuarli** [comment]
> @joshuarli That's a good point, I wonder if that's how users would approach it though? If you're fixing tests it seems likely that you'll look for failed jobs rather than a specific check?

Well - one check is one matrix job right? If you're looking to see if you fixed a specific test and you failed a few, it'll be hard to find that specific test if it's rebalanced. If it's only one test fail then it's easy to click on whatever job fail. Actually, if there's a test that flakes in another job then it's still the same situation. Does this make sense?

**Comment by billyvg** [comment]
Yeah, I'm just assuming you would click all of the failed jobs - but in the case of a flake that would definitely be annoying. We never have flakes though so it's fine _/s_

**Comment by billyvg** [comment]
Can't fully implement this yet because of this bug https://github.com/actions/runner/issues/1578

**Pull request closed without merging**